File: client/lib/modules/documents/game_digest.dart
import 'package:espy/modules/documents/game_entry.dart';

class GameDigest {
  final int id;
  final String name;
  final String? cover;

  final int releaseDate;
  final double rating;

  final List<String> collections;
  final List<String> developers;
  final List<String> publishers;

  GameDigest({
    required this.id,
    required this.name,
    this.cover,
    this.releaseDate = 0,
    this.rating = 0,
    this.collections = const [],
    this.developers = const [],
    this.publishers = const [],
  });

  GameDigest.fromGameEntry(GameEntry gameEntry)
      : this(
          id: gameEntry.id,
          name: gameEntry.name,
          cover: gameEntry.cover?.imageId,
          releaseDate: gameEntry.releaseDate,
          rating: gameEntry.igdbRating,
          collections: [
            for (final collection in gameEntry.collections) collection.name
          ],
          developers: [
            for (final company in gameEntry.developers) company.name
          ],
          publishers: [
            for (final company in gameEntry.publishers) company.name
          ],
        );

  GameDigest.fromJson(Map<String, dynamic> json)
      : this(
          id: json['id']!,
          name: json['name']!,
          cover: json['cover'],
          releaseDate: json['release_date'] ?? 0,
          rating: json['rating'] ?? 0,
          collections: [
            for (final collection in json['collections'] ?? []) collection,
          ],
          developers: [
            for (final company in json['developers'] ?? []) company,
          ],
          publishers: [
            for (final company in json['publishers'] ?? []) company,
          ],
        );

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      if (cover != null) 'cover': cover,
      'release_date': releaseDate,
      if (rating != 0) 'rating': rating,
      if (collections.isNotEmpty) 'collections': collections,
      if (developers.isNotEmpty) 'developers': developers,
      if (publishers.isNotEmpty) 'publishers': publishers,
    };
  }
}

File: server/src/documents/game_digest.rs
use std::collections::HashSet;

use serde::{Deserialize, Serialize};

use super::GameEntry;

#[derive(Serialize, Deserialize, Default, Debug, Clone)]
pub struct GameDigest {
    pub id: u64,
    pub name: String,

    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cover: Option<String>,

    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub release_date: Option<i64>,

    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rating: Option<f64>,

    #[serde(default)]
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub collections: Vec<String>,

    #[serde(default)]
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub developers: Vec<String>,

    #[serde(default)]
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub genres: Vec<String>,

    #[serde(default)]
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub keywords: Vec<String>,
}

impl GameDigest {
    pub fn new(game_entry: GameEntry) -> Self {
        GameDigest {
            id: game_entry.id,
            name: game_entry.name,

            cover: match game_entry.cover {
                Some(cover) => Some(cover.image_id),
                None => None,
            },
            release_date: game_entry.release_date,
            rating: game_entry.igdb_rating,

            collections: game_entry
                .collections
                .into_iter()
                .map(|collection| collection.name)
                .collect::<HashSet<_>>()
                .into_iter()
                .collect(),

            developers: game_entry
                .developers
                .into_iter()
                .map(|company| company.name)
                .collect::<HashSet<_>>()
                .into_iter()
                .collect(),

            genres: match game_entry.steam_data {
                Some(steam_data) => steam_data
                    .genres
                    .into_iter()
                    .map(|genre| genre.description)
                    .collect(),
                None => game_entry.genres,
            },
            keywords: game_entry.keywords,
        }
    }
}
